pkg/config/validator_image_platform.go

```diff
@@ -0,0 +1,62 @@
+package config
+
+import (
+	"fmt"
+	"runtime"
+
+	"github.com/samber/lo"
+)
+
+type imagePlatformValidator struct{}
+
+func newImagePlatformValidator() *imagePlatformValidator {
+	return &imagePlatformValidator{}
+}
+
+func (v *imagePlatformValidator) Validate(rawStapelImages []*rawStapelImage, rawImagesFromDockerfile []*rawImageFromDockerfile) error {
+	// Collect all combinations of image:platform from both sources
+	allImagesPlatforms := make([]lo.Tuple2[string, string], 0, len(rawStapelImages)+len(rawImagesFromDockerfile))
+
+	for _, img := range rawStapelImages {
+		allImagesPlatforms = append(allImagesPlatforms, v.crossJoinImagesPlatforms(img.Images, img.Platform)...)
+	}
+
+	for _, img := range rawImagesFromDockerfile {
+		allImagesPlatforms = append(allImagesPlatforms, v.crossJoinImagesPlatforms(img.Images, img.Platform)...)
+	}
+
+	// Check dependencies for stapel images
+	for _, img := range rawStapelImages {
+		for _, dep := range img.RawDependencies {
+			_, rightDiff := lo.Difference(allImagesPlatforms, v.crossJoinImagesPlatforms([]string{dep.Image}, img.Platform))
+
+			if len(rightDiff) > 0 {
+				return v.newImagePlatformError(img.Images[0], rightDiff[0].A, rightDiff[0].B)
+			}
+		}
+	}
+
+	// Check dependencies for dockerfile images
+	for _, img := range rawImagesFromDockerfile {
+		for _, dep := range img.RawDependencies {
+			_, rightDiff := lo.Difference(allImagesPlatforms, v.crossJoinImagesPlatforms([]string{dep.Image}, img.Platform))
+
+			if len(rightDiff) > 0 {
+				return v.newImagePlatformError(img.Images[0], rightDiff[0].A, rightDiff[0].B)
+			}
+		}
+	}
```

Comment:
It also applies to `from`, `fromImage`, `fromArtifact`, and `import.[from|image]`.